8/n Read-only mode: Input rendering

Ankush-lastmile opened the pull request:
8/n Read-only mode: Input rendering





Attempted to add read-only functionality wherever applicable to input renderings.

## Testplan

| ![image1](https://github.com/lastmile-ai/aiconfig/assets/141073967/31405238-bbd1-42c0-a5fc-58102fe05021) | ![image2](https://github.com/lastmile-ai/aiconfig/assets/141073967/bb8d6543-25d0-4b61-826b-ded9ddab4295) |
| --- | --- |
| ![image1](https://github.com/lastmile-ai/aiconfig/assets/141073967/dcec6c8b-fbe8-4aea-b2f1-f55153866385) | ![image2](https://github.com/lastmile-ai/aiconfig/assets/141073967/8e90c287-3304-484e-8831-496b0f5cd2a8) |
| ![image1](https://github.com/lastmile-ai/aiconfig/assets/141073967/38e2fc27-bb39-4143-8dc1-67876110f1c6) | ![image2](https://github.com/lastmile-ai/aiconfig/assets/141073967/34d4db13-74b5-4617-bb5c-5c851e41afb8) |
| ![image1](https://github.com/lastmile-ai/aiconfig/assets/141073967/80953234-3a0c-4aa5-bfc5-7d7c9176be58) | ![image2](https://github.com/lastmile-ai/aiconfig/assets/141073967/4d66ec78-04f8-45ae-b396-a91e8f3355d3) |

## dependencies

built ontop of #961

Ankush-lastmile commented on python/src/aiconfig/editor/client/src/components/prompt/prompt_input/schema_renderer/PromptInputDataSchemaRenderer.tsx:
```diff
@@ -14,12 +15,15 @@ export default memo(function PromptInputDataSchemaRenderer({
           value={data ? (data as string) : ""}
           onChange={(e) => onChangeData(e.target.value)}
+          disabled={readOnly}
           placeholder="Type a prompt"
```
should this say something else in readonly mode?

rossdanlm replied:
Placeholder is a value for when there is no data already and asking user to do an action. In our case let's set this to "" in readOnly mode (can you test example where the input string is empty) for both readOnly and normal? Thanks!

Ankush-lastmile replied:
placeHolder no longer needed since spoiler will render input which will be an undefined/empty string

Ankush-lastmile commented on python/src/aiconfig/editor/client/src/components/prompt/prompt_input/schema_renderer/PromptInputSchemaRenderer.tsx:
```diff
@@ -77,6 +80,7 @@ export default memo(function PromptInputSchemaRenderer(props: Props) {
         label="Prompt"
         onChange={(e) => props.onChangeInput(e.target.value)}
         placeholder="Type a prompt"
+        disabled = {readOnly}
```
same here

rholinshead replied:
And here :p

rossdanlm replied:
Same about making placeholder empty string for readOnly mode

Ankush-lastmile replied:
readOnly = False
<img width="1085" alt="Screenshot 2024-01-19 at 10 28 40 AM" src="https://github.com/lastmile-ai/aiconfig/assets/141073967/2bb1babd-6799-433c-8873-0d5bd3ffc8d7">

readOnly = True
<img width="990" alt="Screenshot 2024-01-19 at 10 28 54 AM" src="https://github.com/lastmile-ai/aiconfig/assets/141073967/2ea4c77d-d79b-4ece-a864-98b28baed89e">

placeHolder no longer needed since spoiler will render input which will be an undefined/empty string

rholinshead commented on python/src/aiconfig/editor/client/src/components/JSONEditorToggleButton.tsx:
```diff
@@ -10,8 +12,11 @@ export default function JSONEditorToggleButton({
 }: Props) {
+  const { readOnly } = useContext(AIConfigContext);
+
+  const toggleJSONButtonLabel = !readOnly ? "Toggle JSON editor" : "View JSON";
```
nit, prefer flipping so no negation is needed
```
const toggleJSONButtonLabel = readonly ? "View JSON" : "Toggle JSON Editor";

rossdanlm replied:
For context Ankush, I did it the other way for things like `!readOnly ? <addPromptButton>... : undefined` because I think it's clearer to show if the last value is undefined or null, but for this case I agree with Ryan

Ankush-lastmile replied:
updated

rholinshead commented on python/src/aiconfig/editor/client/src/components/prompt/prompt_input/PromptInputConfigRenderer.tsx:
```diff
@@ -11,10 +12,12 @@ export default memo(function PromptInputConfigRenderer({
     <Textarea
       value={input as string}
       onChange={(e) => onChangeInput(e.target.value)}
+      disabled={readOnly}
```
Instead of disabled textarea, let's render the nicer TextRenderer that text outputs use. Maybe we could also wrap that with a Spoiler in case the input is long (see TextInputCellRenderer in lastmile - https://github.com/lastmile-ai/lastmile/blob/bd3b54248682aed2720f7c2c3caefc8ebda335ee/src/components/playground_v2/cell_renderers/TextInputCellRenderer.tsx#L143-L156)

Ankush-lastmile replied:
updated 
readOnly = False
<img width="1055" alt="Screenshot 2024-01-19 at 10 16 06 AM" src="https://github.com/lastmile-ai/aiconfig/assets/141073967/c662e114-8c7c-488f-89e4-ad9e9ac6b460">

readOnly = True
<img width="995" alt="Screenshot 2024-01-19 at 10 17 10 AM" src="https://github.com/lastmile-ai/aiconfig/assets/141073967/110306a1-ec08-4836-a732-186a5e3a91e5">

rholinshead commented on python/src/aiconfig/editor/client/src/components/prompt/prompt_input/schema_renderer/PromptInputAttachmentsSchemaRenderer.tsx:
```diff
@@ -38,7 +40,7 @@ function EditableAttachmentRenderer({
         />
-      ) : (
+      ) : !readOnly && (
         <AttachmentUploader
```
If they have a prompt that takes attachments but it has none, will this just be a blank prompt? Might be better to just render the AttachmentUploader still and have the dropzone disabled to denote that it's an input for attachments?

rholinshead replied:
I don't feel too strongly here -- can you compare how they look?

Ankush-lastmile replied:
<img width="975" alt="Screenshot 2024-01-19 at 10 19 56 AM" src="https://github.com/lastmile-ai/aiconfig/assets/141073967/e3cee944-0267-45bc-915c-723972935156">

disabled doesn't change how its rendered but only disables the dropzone. 
<img width="986" alt="Screenshot 2024-01-19 at 10 22 55 AM" src="https://github.com/lastmile-ai/aiconfig/assets/141073967/42d6e35a-e583-4ff2-bba6-da62b2ff68f1">

rholinshead replied:
I think the disabled dropzone is better than blank for now, otherwise there's no indication of what the input is 'supposed to be'. We do the same in lastmile so it is consistent as well

Ankush-lastmile replied:
updated

rholinshead commented on python/src/aiconfig/editor/client/src/components/prompt/prompt_input/schema_renderer/PromptInputDataSchemaRenderer.tsx:
```diff
@@ -14,12 +15,15 @@ export default memo(function PromptInputDataSchemaRenderer({
         <Textarea
           value={data ? (data as string) : ""}
           onChange={(e) => onChangeData(e.target.value)}
+          disabled={readOnly}
```
Same here r.e. using the TextRenderer w/ Spoiler

Ankush-lastmile replied:
updated

rossdanlm commented on python/src/aiconfig/editor/client/src/components/JSONRenderer.tsx:
```diff
@@ -14,7 +15,9 @@ export default memo(function JSONRenderer({
 }: Props) {
-  return !onChange ? (
+  const { readOnly } = useContext(AIConfigContext);
+  // Prism is a read only renderer.
```
Nice comment!

Ankush-lastmile replied:
![](https://media0.giphy.com/media/v1.Y2lkPTc5MGI3NjExNHFuY3VwdW54NXRuOHczdDV0bjhwZDQzMmxjMjE0cGN5M2oxcXdudSZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/TilmLMmWrRYYHjLfub/giphy.gif)

rossdanlm reviewed (approved):
Accepting to unblock but pls land with Ryan and my comments!

Ankush-lastmile commented:
- remove readOnly on attachment renderer

Ankush-lastmile commented:
> Accepting to unblock but pls land with Ryan and my comments!

Addressed all comments, Shipping this diff. Will fix forward if anything else might have been missed
